### DISCUSSION TITLE
Plotting meshes with different value ranges

### DISCUSSION THREAD

**Discussion by AlexanderJuestel** [pyvista/pyvista#3899] [Q&A] [answered]
Hello, 

I would like to plot seismic data (confidential, unfortunately) in PyVista but the seismic data has different value ranges. I would like PyVista now to scale the colorbars for both meshes to their respective extreme values, so that they look the same. Currently, the colors are scaled to seismic data with the larger value range and the one with the low values appears very weak. I tried using the  `clim` keyword/parameter but it only works on the larger value range. Also changing the order of adding the meshes does not have any effect.

![image](https://user-images.githubusercontent.com/45469915/215318760-a43e5f88-66fd-4a59-bcdc-ecfacff2ed62.png)

```ruby
import pyvista as pv
sargs = dict(color='black')

p = pv.Plotter(notebook=True)

p.add_mesh(seismic _large_range, cmap=cmap_seismic, lighting=False, clim=[-50,50], scalar_bar_args=sargs)
p.add_mesh(seismic_small_range, cmap=cmap_seismic, lighting=False, clim=[-5,5])

p.show_grid(color='black')
p.set_background(color='white')
p.set_scale(1,1,3)
p.show()

```

Seismic Data with large value ranges
![image](https://user-images.githubusercontent.com/45469915/215318446-8f11f959-5918-47a6-a7ec-5499f788d0d5.png)

Seismic with small value ranges
![image](https://user-images.githubusercontent.com/45469915/215318459-1bbbcf61-af6f-4072-bb7e-451c04e0887f.png)

**Comment by banesullivan** [comment, marked as answer]
It sounds like the values are fundamentally different for the two meshes and thus, you should not call both of the `"values"` as they will need to be mapped separately.

One array has a range of -61 to 62, while the other -5 to 5. You cannot nicely put these on one scalar bar with a shared clim (assuming even distributions throughout their ranges which looks to be the case).

It sounds like you just need to rename one of the arrays and map them separately:

```py
import pyvista as pv
import numpy as np

a = pv.Cone(center=(0, 0, 0))
b = pv.Cube(center=(1, 0, 0))

a['values_a'] = np.linspace(-60, 60, a.n_cells)
b['values_b'] = np.linspace(-6, 6, b.n_cells)

p = pv.Plotter()
p.add_mesh(a)
p.add_mesh(b)
p.show()
```
![screenshot](https://user-images.githubusercontent.com/22067021/215345047-4e9cbc98-36ef-4577-b74b-9f23de25d257.png)

**Reply by AlexanderJuestel** [reply]
Hey @banesullivan, 

okay, so they are mapped together. I knew that there is an easy solution to that. Luckily, I only have three datasets with multiple lines for each dataset.

Thanks for the quick help. I also hope that my questions lately will benefit any other PyVista users as well :) 

Cheers Alex

**Comment by AlexanderJuestel** [comment]
Hey @banesullivan,

I have encountered a similar issue using the Multi Window Plot (https://docs.pyvista.org/version/stable/examples/02-plot/multi-window.html). I have different data arrays but all meshes still scale to the last mesh/data array even though I am calling different arrays. 

![image](https://user-images.githubusercontent.com/45469915/236194122-22fcf61b-09c5-4232-a8e3-8f91b209a523.png)

![image](https://user-images.githubusercontent.com/45469915/236193750-a4bb819f-614b-4ed8-9fa8-c98c2ddb67d5.png)

```ruby
p = pv.Plotter(shape=(3,2), notebook=False)

p.subplot(0, 0)
p.add_text("Depth [m]", font_size=16)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0], scalars='Depth [m]',  clim=[-4000,0], opacity=0.7)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0].contour(), color='white', opacity=0.7)
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,2)

p.subplot(0, 1)
p.add_text("Pore Pressure [MPa]", font_size=16)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0], scalars='Pore Pressure [MPa]', opacity=0.7, clim=[0,40])
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0].contour(), color='white', opacity=0.7)
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,2)

p.subplot(1, 0)
p.add_text("Vertical Stress [MPa]", font_size=16)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0], scalars='Vertical Stress [MPa]', opacity=0.7, clim=[0,110])
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0].contour(), color='white', opacity=0.7)
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,2)

p.subplot(1, 1)
p.add_text("Dips [°]", font_size=16)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0], scalars='Dip [°]', opacity=0.7, clim=[0,9])
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,2)

p.subplot(2,0)
p.add_text("Azimuth [°]", font_size=16)
p.add_mesh(mesh=dict_surfaces['Weisweiler Sprung'][0], scalars='Azimuth [°]', opacity=0.7, clim=[0,360])
p.set_background('white')
p.show_grid(color='black')
p.set_scale(1,1,2)

# Display the window
p.show()
```

**Reply by banesullivan** [reply]
The problem is you are using the same mesh across multiple plotters and activating a scalar on that mesh in multiple places. This issue is tracked in https://github.com/pyvista/pyvista/issues/542. The best workaround is to create a copy of the mesh for each subplot (fairly inexpensive really):

```py
import pyvista as pv

grid = pv.Wavelet()
grid["RTData**2"] = grid["RTData"] ** 2

pl = pv.Plotter(shape=(1,2))
pl.subplot(0,0)
pl.add_mesh(grid, scalars="RTData")
pl.subplot(0,1)
pl.add_mesh(grid.copy(), scalars="RTData**2")
pl.show()
```

Or you could use some more advanced features in PyVista with VTK algorithms and configure the active scalars algorithm in a pipeline for each of these. Here is an example:

```py
import pyvista as pv
from pyvista.utilities.algorithms import active_scalars_algorithm

grid = pv.Wavelet()
grid["RTData**2"] = grid["RTData"] ** 2

pl = pv.Plotter(shape=(1,2))
pl.subplot(0,0)
pl.add_mesh(active_scalars_algorithm(grid, "RTData"))
pl.subplot(0,1)
pl.add_mesh(active_scalars_algorithm(grid, "RTData**2"))
pl.show()
```

**Reply by banesullivan** [reply]
I've been meaning to just use `active_scalars_algorithm` directly in `add_mesh()` as it would completely solve this issue behind the scenes

**Reply by AlexanderJuestel** [reply]
That certainly solved the issue! Also thanks for linking the actual issue!

![image](https://user-images.githubusercontent.com/45469915/236421726-31580063-d7cb-4f01-8f6a-f84f1b4ae4b8.png)
